src/holder.rs

```diff
@@ -1,7 +1,79 @@
 use crate::sync::atomic::AtomicPtr;
 use crate::{Domain, HazPtrObject, HazPtrRecord};
+use std::mem::{ManuallyDrop, MaybeUninit};
 use std::sync::atomic::Ordering;
 
+pub struct HazardPointerArray<'domain, F, const N: usize> {
+    haz_ptrs: [ManuallyDrop<HazardPointer<'domain, F>>; N],
+}
+
+impl<'domain, F, const N: usize> HazardPointerArray<'domain, F, N> {
+    pub fn hazard_pointers<'array>(&'array mut self) -> [&'array mut HazardPointer<'domain, F>; N] {
+        let mut out: [MaybeUninit<&'array mut HazardPointer<'domain, F>>; N] =
+            [(); N].map(|_| MaybeUninit::uninit());
+
+        for (i, hazptr) in self.haz_ptrs.iter_mut().enumerate() {
+            out[i].write(hazptr);
+        }
+
+        //
+        // # Safety
+        //
+        // We have initialized every element of the array with our for loop above
```

Comment:
I think this is insufficient. You also need to guarantee here that the returned lifetime constrained the references sufficiently that they remain valid for as long as is necessary. Which we guarantee with `'array` being used on both `&mut self` and the returned `&mut`s. (All that's needed is to add to the comment).

Reply:
I don't think this is true. The `assume_init()` only assumes that the MaybeUnint has been initialized. Whether the lifetime is valid or not is not related to this unsafe block. The compiler would error at the `out[i].write(hazptr);` if there was a lifetime mismatch.

Further, the input &mut and the output [&mut; N] are the only borrows in this function, meaning rust's lifetime elision can produce the correct lifetimes without any manual specifying.

Reply:
I suppose if the `out[i].write(hazptr)` was not present, it would be possible to assume_init() a return value with an invalid lifetime, but if that were the case not only would the lifetime be invalid but the `we have initialized every element` comment would also be invalid

Reply:
Good point, I agree, thanks for the thorough description :+1: